### PR TITLE
Solve trigonometric / hyperbolic equations by "inverting"

### LINKED ISSUE #9825: solveset fails on solveset(Eq(tan(x), y), x, domain=S.Reals)
`>>> solveset(Eq(tan(x), y), x, domain=S.Reals)`
`EmptySet()`

However, `solveset(Eq(tan(x), y), x)` raises `NotImplementedError` since inversion is not handled.

### LINKED ISSUE #17334: solveset: current state of _solve_trig and _invert
**[I have slightly updated this post.]**

This is a (rather long) summary/discussion issue that outlines some *specific* shortcomings I noticed while reviewing solveset's capabilities for solving trigonometric equations. (As such it's a complement to/part of https://github.com/sympy/sympy/issues/12340. See also: https://github.com/sympy/sympy/issues/13084)

What bothers me most is that all `solveset(trig equation)` calls always ends up in a substitution based solver, either `_solve_trig1` or `_solve_trig2`. Both solvers have in common that they completely lose the information what the actual trig function was (sin? cos? etc.)

**[fixed for solveset_real]** Here's an example for which `.rewrite(exp)` is used (by `_solve_trig1`):
```
In [19]: solveset(sin(x) - sin(1), x, S.Reals)
Out[19]: Union(Intersection(ImageSet(Lambda(_n, -I*(I*(2*_n*pi + atan(sin(1)/cos(1))) + log(sqrt(cos(1)**2 + sin(1)**2)))), Integers), Real
s), Intersection(ImageSet(Lambda(_n, -I*(I*(2*_n*pi - atan(sin(1)/cos(1)) + pi) + log(sqrt(cos(1)**2 + sin(1)**2)))), Integers), Reals))

In [20]: solveset(sin(x) - sqrt(5)/3, x, S.Reals)
Out[20]: Union(ImageSet(Lambda(_n, 2*_n*pi + atan(sqrt(5)/2)), Integers), ImageSet(Lambda(_n, 2*_n*pi - atan(sqrt(5)/2) + pi), Integers))
```
(Sorry, unicode doesn't always work out fine for copy&paste.)
It's clearly noticeable that an exponential equation has in fact been solved, the determination of complex arguments being responsible for the `atan`s. The first example can be properly simplified using `simplify`, but that's not always the case! Just try: `solveset(tan(x)-tan(pi/5),x,S.Reals)`...

Yet, there's a powerful alternative solver for these basic equations: `_invert`. Compare the above with:
```
In [25]: from sympy.solvers.solveset import _invert

In [26]: _invert(sin(x), sin(1), x, S.Reals)
Out[26]: (x, ImageSet(Lambda(_n, (-1)**_n + _n*pi), Integers))

In [27]: _invert(sin(x), sqrt(5)/3, x, S.Reals)
Out[27]: (x, ImageSet(Lambda(_n, (-1)**_n*asin(sqrt(5)/3) + _n*pi), Integers))

In [28]: _invert(tan(x), tan(pi/5), x, S.Reals)
Out[28]: (x, ImageSet(Lambda(_n, _n*pi + pi/5), Integers))
```

Don't get me wrong: `_solve_trig1` and `_solve_trig2` are very important parts of solveset, they're fundamental for solving equations with *mixed* trig functions. But for equations that are amenable to a more direct approach, they're hardly optimal.

Currently `_invert` is already called upon by other solvers, but not by `_solve_trig`! (Neither by `solve_decomposition` nor `nonlinsolve`)

Examples:
```
In [34]: from sympy.solvers.solveset import _solve_as_poly, _solve_as_rational

In [35]: _solve_as_poly(tan(x)-tan(pi/5), x, S.Reals)
Out[35]: ImageSet(Lambda(_n, _n*pi + pi/5), Integers)

In [36]: _solve_as_rational(tan(x)-tan(pi/5), x, S.Reals)
Out[36]: ImageSet(Lambda(_n, _n*pi + pi/5), Integers)
```
Nice and easy.

In essence: solvers calling `solveset` for trig equations will often produce unusable results for simple equations, while solvers internally calling `_invert` fare better for these types of situation. That's unfortunate.

**1st conclusion: some trig equations *require* `_invert`**

Preliminary local testing shows that calling `_invert` in `_solve_trig` for equations containing only a single trig instance is indeed beneficial! (Fixes or improves upon https://github.com/sympy/sympy/issues/11218, https://github.com/sympy/sympy/issues/12340)

But there's a catch: `_invert` isn't perfect either. Examples:
```
In [41]: _invert(sin(x), 1, x, S.Reals)
Out[41]: (x, ImageSet(Lambda(_n, (-1)**_n*pi/2 + _n*pi), Integers))
```
**[fixed for solveset_real]** `_invert_real` tries to be too "clever" with its use of (-1)**n for sin and csc.
```
In [42]: _invert(sinh(x), 1, x, S.Reals)
Out[42]: (sinh(x), {1})
```
**[fixed for solveset_real]** Hyperbolic functions are not implemented, possibly because `_solve_trig` is supposed to handle all hyperbolic equations when called via `solveset`, but currently:
```
In [43]: solveset(sinh(x)-1, x, S.Reals)
Out[43]: ConditionSet(x, Eq((exp(2*x) - 2*exp(x) - 1)*exp(-x)/2, 0), Reals)
```
not useful...

`_invert_complex` has no code at all for handling trigonometric or hyperbolic functions, yet other solvers do call `_invert` if they're working with substitutions. The following is returning a `ConditionSet` because of this missing support for hyperbolic functions in `_invert_complex`:
```
In [1]: from sympy.solvers.solveset import _solve_as_poly, _solve_as_rational

In [2]: eq = expand((x + 1)**2).subs(x, sinh(x))

In [3]: _solve_as_poly(eq, x, S.Complexes)
```

**2nd conclusion: make `_invert_real` and `_invert_complex` as powerful as possible**

So if nobody tells me otherwise, I'll work on the following:
- [x] improve _invert_real
- [ ] improve _invert_complex (trig)
- [ ] improve _invert_complex (hyperbolic)
- [ ] call _invert in _solve_trig when appropriate (e.g. single trig instance)
- [ ] test cases

### LINKED ISSUE #20798: Solveset returns empty set for cos(2x) = 1/2
Running in https://live.sympy.org/, where the version seems to be 1.5.1 you get the following:
```
>>> solveset(cos(2*x) - 1/2, x, Interval(0, 2*pi))
{π6,5π6,7π6,11π6}
```

However, running 1.7.1, you get:
```
>>> solveset(cos(2*x) - 1/2, x, Interval(0, 2*pi))
∅
```

Any idea where to begin looking for a solution?

### LINKED ISSUE #21296: solveset returns very complicated answers with tan and irrational numbers:
Also solveset returns very complicated answers with tan and irrational numbers:
```julia
In [17]: solveset(tan(x)-S.Half, x, Interval(0, pi/2))       
Out[17]: {atan(1/2)}

In [18]: solveset(tan(x)-sqrt(2), x, Interval(0, pi/2))                                                                                        
Out[18]: 
⎧   ⎛                                                       ⎛   ⎛atan(2⋅√2)⎞⎞⎞⎫
⎪   ⎜   ⎛    _____________________________________⎞         ⎜cos⎜──────────⎟⎟⎟⎪
⎪   ⎜   ⎜   ╱    2⎛atan(2⋅√2)⎞      2⎛atan(2⋅√2)⎞ ⎟         ⎜   ⎝    2     ⎠⎟⎟⎪
⎨-ⅈ⋅⎜log⎜  ╱  sin ⎜──────────⎟ + cos ⎜──────────⎟ ⎟ + ⅈ⋅atan⎜───────────────⎟⎟⎬
⎪   ⎜   ⎝╲╱       ⎝    2     ⎠       ⎝    2     ⎠ ⎠         ⎜   ⎛atan(2⋅√2)⎞⎟⎟⎪
⎪   ⎜                                                       ⎜sin⎜──────────⎟⎟⎟⎪
⎩   ⎝                                                       ⎝   ⎝    2     ⎠⎠⎠⎭

In [19]: solveset(tan(x)-pi, x, Interval(0, pi/2))                                                                                             
Out[19]: 
⎧   ⎛   ⎛            ________________________________________________________________________________________________________________________
⎪   ⎜   ⎜           ╱                                                ⎛    ⎛       2⋅π       ⎞⎞                                               
⎪   ⎜   ⎜          ╱                                                 ⎜atan⎜─────────────────⎟⎟                                               
⎪   ⎜   ⎜         ╱                                                  ⎜    ⎜              4  ⎟⎟                                               
⎪   ⎜   ⎜        ╱            __________________________________     ⎜    ⎜    1        π   ⎟⎟           __________________________________  
⎪   ⎜   ⎜       ╱            ╱                                2      ⎜    ⎜- ────── + ──────⎟⎟          ╱                                2   
⎪   ⎜   ⎜      ╱            ╱        2     ⎛              2  ⎞       ⎜    ⎜       2        2⎟⎟         ╱        2     ⎛              2  ⎞    
⎨   ⎜   ⎜     ╱            ╱      4⋅π      ⎜    1        π   ⎟      2⎜    ⎝  1 + π    1 + π ⎠⎟        ╱      4⋅π      ⎜    1        π   ⎟    
⎪-ⅈ⋅⎜log⎜    ╱            ╱    ───────── + ⎜- ────── + ──────⎟  ⋅sin ⎜───────────────────────⎟ +     ╱    ───────── + ⎜- ────── + ──────⎟  ⋅c
⎪   ⎜   ⎜   ╱            ╱             2   ⎜       2        2⎟       ⎝           2           ⎠      ╱             2   ⎜       2        2⎟    
⎪   ⎜   ⎜  ╱            ╱      ⎛     2⎞    ⎝  1 + π    1 + π ⎠                                     ╱      ⎛     2⎞    ⎝  1 + π    1 + π ⎠    
⎪   ⎜   ⎝╲╱           ╲╱       ⎝1 + π ⎠                                                          ╲╱       ⎝1 + π ⎠                           
⎪   ⎜                                                                                                                                        
⎪   ⎜                                                                                                                                        
⎩   ⎝                                                                                                                                        

_____________________________⎞                            ⎞⎫
   ⎛    ⎛       2⋅π       ⎞⎞ ⎟                            ⎟⎪
   ⎜atan⎜─────────────────⎟⎟ ⎟         ⎛   ⎛    ⎛ 2⋅π  ⎞⎞⎞⎟⎪
   ⎜    ⎜              4  ⎟⎟ ⎟         ⎜   ⎜atan⎜──────⎟⎟⎟⎟⎪
   ⎜    ⎜    1        π   ⎟⎟ ⎟         ⎜   ⎜    ⎜     2⎟⎟⎟⎟⎪
   ⎜    ⎜- ────── + ──────⎟⎟ ⎟         ⎜   ⎜    ⎝1 - π ⎠⎟⎟⎟⎪
   ⎜    ⎜       2        2⎟⎟ ⎟         ⎜cos⎜────────────⎟⎟⎟⎪
  2⎜    ⎝  1 + π    1 + π ⎠⎟ ⎟         ⎜   ⎝     2      ⎠⎟⎟⎬
os ⎜───────────────────────⎟ ⎟ - ⅈ⋅atan⎜─────────────────⎟⎟⎪
   ⎝           2           ⎠ ⎟         ⎜   ⎛    ⎛ 2⋅π  ⎞⎞⎟⎟⎪
                             ⎟         ⎜   ⎜atan⎜──────⎟⎟⎟⎟⎪
                             ⎠         ⎜   ⎜    ⎜     2⎟⎟⎟⎟⎪
                                       ⎜   ⎜    ⎝1 - π ⎠⎟⎟⎟⎪
                                       ⎜sin⎜────────────⎟⎟⎟⎪
                                       ⎝   ⎝     2      ⎠⎠⎠⎭
```

Related ISSUE: #17667
_Originally posted by @oscarbenjamin in https://github.com/sympy/sympy/issues/17667#issuecomment-536329206_

### PR DESCRIPTION
**Pull request by gschintgen**
#### References to other Issues or PRs

Fixes #21296
Improves #17667
Improves #17579
Fixes #17334
Related to #12340
Closes #20798
Fixes #9825

#### Brief description of what is fixed or changed
This PR adds
- inverting of complex trigonometric functions
- inverting of complex hyperbolic functions
- domain checking to the inverter for real-valued trig
- a lot of tests

It also improves `_solve_trig` by using the inverter for more types of equations. It adds a bit of domain checking.

#### Other comments
The downside of the more precise results with this PR is that there are now more `ConditionSet`s when symbols are involved.

In order to make the test suite pass, I also had to fix two bugs in `nonlinsolve`. Even with the debugger these were quite frustrating to track down. (See the relevant commits.)

I guess there's always more to be done, but at this point I'd just like to kindly request a review in order to get this merged. (I'll remove the draft status as soon as the tests have successfully completed.)

Further work that could be done: unify trig/hyp inverting a bit more, refactor that part a bit.

#### Release Notes

* solvers
  * Implemented `invert`ing of complex-valued trigonometric or hyperbolic functions.
  * Improved `solveset` by using the inverter more often for trigonometric or hyperbolic equations.
  * Added domain checking to the real-valued trigonometric inverter used by `solveset`.
  * nonlinsolve: Fix two bugs related to solving systems involving trigonometric functions. (Wrong results fixed.)

### CODE REVIEW

File: sympy/solvers/solveset.py [hunk 1]
```diff
@@ -394,7 +407,50 @@ def _invert_trig_hyp_real(f, g_ys, symbol):
...
+            cot : ((acot,), pi, S.Reals),
+            sec : ((asec, lambda y: -asec(y)), 2*pi,
+                   Union(Interval(-oo, -1), Interval(1, oo))),
+            csc : ((acsc, lambda y: pi-acsc(y)), 2*pi,
+                   Union(Interval(-oo, -1), Interval(1, oo)))}
+
+        retset = S.EmptySet
+        for g in g_ys:
+            retset += create_return_set(*trig_inverses[f.func])
+        return (symbol, retset)
```
[the remark this answers is not in the document]
Comment (gschintgen) [reply sympy/solvers/solveset.py]: Now that you're pointing it out...
I've updated the PR.

**Comment by oscarbenjamin** [comment]
This looks good to me.

File: sympy/solvers/solveset.py [hunk 2]
```diff
@@ -394,7 +407,47 @@ def _invert_trig_hyp_real(f, g_ys, symbol):
...
-        # XXX: a domain check must be added.
+        trig_inverses = {
+            sin : ((asin, lambda y: pi-asin(y)), 2*pi, Interval(-1, 1)),
+            cos : ((acos, lambda y: -acos(y)), 2*pi, Interval(-1, 1)),
+            tan : ((atan,), pi, S.Reals),
+            cot : ((acot,), pi, S.Reals),
+            sec : ((asec, lambda y: -asec(y)), 2*pi,
+                   Union(Interval(-oo, -1), Interval(1, oo))),
+            csc : ((acsc, lambda y: pi-acsc(y)), 2*pi,
+                   Union(Interval(-oo, -1), Interval(1, oo)))}
```
[the remark this answers is not in the document]
Comment (gschintgen) [reply sympy/solvers/solveset.py:418]: Oh, that's expensive. Thanks for the tip. I'll push the changes in a few seconds.

**Comment by gschintgen** [comment]
Thanks!

**Pull request merged**